cmd/backup/main.go

```diff
@@ -4,19 +4,28 @@
 package main
 
 import (
+	"flag"
 	"fmt"
+	"log"
 	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/robfig/cron/v3"
 )
 
-func main() {
-	s, err := newScript()
+func runBackup(c *Config) {
```

Comment:
I guess this function will now have to return an error, so that we can:
- exit 1 when not running in the foreground
- log/notify, but not exit when running in the foreground

Reply:
I tried to wrap my head around how `defer` and all that works in Go. I think there was a bug in the code regarding the lock, we instantly unlocked it in the next statement. 
> [The deferred call's arguments are evaluated immediately, but the function call is not executed until the surrounding function returns.](https://go.dev/tour/flowcontrol/12)

I will go and do some restructure for the rest of the codebase by removing any reference to `panic` ( and the related `must` function. )

Do you have any rule on how to split what is sent into the script logger ( `s.logger` ) and what is fine using the global `log` object? ( I believe the only different is that the first one will provide some stats regarding individual script runs? )

Reply:
> I tried to wrap my head around how defer and all that works in Go. I think there was a bug in the code regarding the lock, we instantly unlocked it in the next statement.

Very good catch, seems this was broken here https://github.com/offen/docker-volume-backup/pull/263/files and I did not notice in review. I'll patch this directly as it's kind of bad.

> Do you have any rule on how to split what is sent into the script logger

The current setup is mostly a remnant of a time when the standard library did not have structured logging and package `logrus` was used (changed in https://github.com/offen/docker-volume-backup/pull/247).

Currently, the idea is to always use the logger instance (because this is how logrus works), but I am not sure if this is worth it. The nice thing about doing it like this is obviously testability, then this repo doesn't have any unit tests and only integration tests (it would be nice to have unit tests though, but I simply never find the time to add them). In any case I think we should pick one of the two approaches and stick to it always, i.e. always use the logger instance or always use the global one. Is there anything that makes it especially hard to always use the logger instance in your changeset?

Reply:
Fixing the locking in https://github.com/offen/docker-volume-backup/pull/339

Reply:
> Is there anything that makes it especially hard to always use the logger instance in your changeset?

Currently the logger instance is bound to the `script` object, which is used for one execution. This is also used to collect and provide logs to the notifications. This feature demands the use of an individual logger instance for every backup run. 

My idea to stay consistent through the whole application:
- On startup create the logger that will be used through the lifetime of the application.
- Inject this into every backup run, so they can wrap around it and provide both logs to stdout and into the buffer.

I think this would result in a consistent experience ( every log message is done via `logger.Info` ) and keep the same functionality.

Reply:
Sounds good to me :+1:

Reply:
I decided to make #341  where the logging could be standardized and improved in the whole application. Right now the main loop will use the logger initialized at startup.